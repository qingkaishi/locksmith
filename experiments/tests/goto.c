#include <pthread.h>
#include <stdio.h>

int *j;
pthread_mutex_t l;

void * dostuff(void * a) {
  int k;
  
  pthread_mutex_lock(&l);
  j = &k;
  pthread_mutex_unlock(&l);
  return NULL;
}

int main() {
  pthread_t t1;
  pthread_mutex_init(&l, NULL);
  int i = 1;
  pthread_create(&t1, NULL, dostuff, NULL);
  //spaghetti starts:
  label1:
    if (i) goto label5;
    if (!i) goto label6;
  label2:
    goto label3;
  label4:
  label3:
    return 1;
  label5:
    goto label4;
  label6:
  j = &i;  //  race
  goto label5;
}


