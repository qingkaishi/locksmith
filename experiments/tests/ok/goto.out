************************* STARTING *************************

shared: &j:goto.c:4

Warning: Possible data race: &j:goto.c:4 is not protected!
 references:
  dereference of &j:goto.c:4 at goto.c:11
    &j:goto.c:4
  locks acquired:
    concrete l:goto.c:18
    l:goto.c:5
  in: main at goto.c:16 

  dereference of &j:goto.c:4 at goto.c:33
    &j:goto.c:4
  locks acquired:
    <empty>
  in: main at goto.c:16 


*************************** DONE ***************************
